lnwire/update_add_htlc.go

```diff
@@ -152,16 +185,51 @@ func (c *UpdateAddHTLC) Encode(w *bytes.Buffer, pver uint32) error {
 		return err
 	}
 
-	// Only include blinding point in extra data if present.
+	// Construct a slice of all the records that we should include in the
+	// extra data field.
 	var records []tlv.RecordProducer
 
+	// Only include blinding point in extra data if present.
 	c.BlindingPoint.WhenSome(func(b tlv.RecordT[BlindingPointTlvType,
 		*btcec.PublicKey]) {
 
 		records = append(records, &b)
 	})
 
-	err := EncodeMessageExtraData(&c.ExtraData, records...)
+	// Extend the 'records' slice with TLV records from the custom records
+	// field.
+	//
+	// As a sanity check, we will first ensure that the existing records
+	// TLV types are less than the minimum custom records TLV type.
+	for _, recordProducer := range records {
+		record := recordProducer.Record()
+		recordTlvType := uint64(record.Type())
+
+		if recordTlvType >= MinCustomRecordsTlvType {
+			return fmt.Errorf("records contain TLV type that is "+
+				"unexpectedly greater than or equal to the "+
+				"minimum TLV type of custom records: %d",
+				recordTlvType)
+		}
+	}
+
+	// Validate custom records field. This will ensure that all custom
+	// records are within the correct TLV type range.
+	err := c.CustomRecords.Validate()
```

Comment:
Combined with my comment above, maybe we could extract all of this into a more generic function.
Something like this:

```go
func MergeRecordsWithCustom(specRecords []tlv.RecordProducer,
	customRecords []tlv.RecordProducer) ([]tlv.RecordProducer, error) {

	// As a sanity check, we will first ensure that the spec compliant
	// record TLV types are in the correct range (below the custom record
	// start value).
	for _, specRecord := range specRecords {
		record := specRecord.Record()
		recordTlvType := uint64(record.Type())

		if recordTlvType >= MinCustomRecordsTlvType {
			return nil, fmt.Errorf("spec compliant record TLV "+
				"type is unexpectedly greater than or equal "+
				"to the minimum TLV type of custom records: %d",
				recordTlvType)
		}
	}

	// Convert the custom records map to TLV records and append them to the
	// exiting records slice.
	for _, customRecord := range customRecords {
		record := customRecord.Record()
		recordTlvType := uint64(record.Type())

		if recordTlvType < MinCustomRecordsTlvType {
			return nil, fmt.Errorf("custom records entry with TLV "+
				"type below minimum value for custom "+
				"records: %d", MinCustomRecordsTlvType)
		}
	}

	return specRecords, nil
}
```

Then this code section here can just be:

```go
	// Construct a slice of all the records that we should include in the
	// extra data field.
	var specRecords []tlv.RecordProducer

	// Only include blinding point in extra data if present.
	c.BlindingPoint.WhenSome(func(b tlv.RecordT[BlindingPointTlvType,
		*btcec.PublicKey]) {

		specRecords = append(specRecords, &b)
	})

	// Convert the custom records map to TLV record producers.
	customRecords := fn.Map(func(a tlv.Record) tlv.RecordProducer {
		return &recordProducer{a}
	}, tlv.MapToRecords(c.CustomRecords))

	records, err := MergeRecordsWithCustom(specRecords, customRecords)
	if err != nil {
		return err
	}
```

And `CustomRecords` would just be `tlv.TypeMap`.